sched/task/task_tls_alloc.c

```diff
@@ -0,0 +1,113 @@
+/****************************************************************************
+ * sched/task/task_tls_alloc.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <sched.h>
+#include <errno.h>
+
+#include <nuttx/tls.h>
+#include <nuttx/semaphore.h>
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+#if CONFIG_TLS_TASK_NELEM > 0
+
+static int g_tlsset = 0;
+static sem_t g_tlssem = SEM_INITIALIZER(1);
+static tls_dtor_t g_tlsdtor[CONFIG_TLS_TASK_NELEM];
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int task_tls_alloc(tls_dtor_t dtor)
+{
+  int ret;
+  int candidate;
+
+  ret = nxsem_wait(&g_tlssem);
+
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = -EAGAIN;
+
+  for (candidate = 0; candidate < CONFIG_TLS_TASK_NELEM; candidate++)
+    {
+      int mask = 1 << candidate;
+      if ((g_tlsset & mask) == 0)
+        {
+          g_tlsset |= mask;
+          g_tlsdtor[candidate] = dtor;
+          ret = candidate;
+          break;
+        }
+    }
+
+  nxsem_post(&g_tlssem);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: task_tls_destruct
+ *
+ * Description:
+ *   Destruct all TLS data element associated with allocated key
+ *
+ * Input Parameters:
+ *    None
+ *
+ * Returned Value:
+ *   A set of allocated TLS index
+ *
+ ****************************************************************************/
+
+void task_tls_destruct(void)
```

Comment:
I would like to get some more explanation about operation of this API. Maybe I'm missing something and that brings my concerns up.
For example `CONFIG_TLS_TASK_NELEM` is configured to `3`. Then each task (`struct task_info_s`) will be equipped with additional `3` elements of `uintptr_t` while globally we can have only `CONFIG_TLS_TASK_NELEM` number of destructors (`static tls_dtor_t g_tlsdtor[CONFIG_TLS_TASK_NELEM];`). So if I'm having `10` tasks the `3 * 10 * sizeof(uintptr_t)` will be additionally allocated for tasks and `3 * sizeof(tls_dtor_t)` for destructors. For example `task_A` call `task_tls_alloc()` one time and `task_B` call `task_tls_alloc()` two times. Then `task_A` exits and `task_tls_destruct` is called. The loop will iterate `3` times and globally `if ((g_tlsset & mask) != 0)` will always hit a `true` condition (one time for `task_A` and two times for `task_B`) and `dtor((void *)elem);` will be called `3` times passing 2 times `(void *)0` to `dtor()`. So the questions are next:
1. Is it expected that `task_tls_destruct()` will call TLS destructors for other tasks?
2. If yes, then should we have a prerequisite that `dtor()` should check for `NULL` inside that call?
3. Is it expected that after `task_tls_destruct()` is called the while all `CONFIG_TLS_TASK_NELEM` are allocated the new `task_tls_alloc()` will always fail because `g_tlsset` never resets bits?
4. What it `task_tls_destruct()` is called in the middle of `task_tls_alloc()`, for example after `g_tlsdtor[candidate] = dtor;` and before `ret = candidate;`? Will the `dtor` be ready to accept `NULL` (that actually comes back to 2)

Reply:
These APIs is very similar to pthread_key_create series APIs, the only different is it works at task level, then:
> 1. Is it expected that task_tls_destruct() will call TLS destructors for other tasks?

Yes, if any tls key allocated, we must call task_tls_destruct for all tasks since scheduler don't know wether was the task specific context used.

> If yes, then should we have a prerequisite that dtor() should check for NULL inside that call?

Yes, we can check the data element as pthread does.

> 3. Is it expected that after task_tls_destruct() is called the while all CONFIG_TLS_TASK_NELEM are allocated the new task_tls_alloc() will always fail because g_tlsset never resets bits?

Yes in current implementation, maybe we can add a reference counter for each bit to recycle it ? Or maybe it's not a problem since in current workload, only few application or midlleware like libuv need it, and in such a complex system, the counter is hard to get chance to return zero.

> 4. What it task_tls_destruct() is called in the middle of task_tls_alloc(), for example after g_tlsdtor[candidate] = dtor; and before ret = candidate;? Will the dtor be ready to accept NULL (that actually comes back to 2)

It's safe in this situation since task_tls_destruct and task_tls_alloc is independent in different task, if task A use relative API (task_tls_alloc), it's dtor is used for all tasks but it's data element is only used in task A.

> 5. Is it expected to have 3 * (N - 1) * sizeof(uintptr_t) (where N is the number of tasks in the system) of waste memory

Yes, so these APIs disabled by default.

Reply:
> I would like to get some more explanation about operation of this API. Maybe I'm missing something and that brings my concerns up. For example `CONFIG_TLS_TASK_NELEM` is configured to `3`. Then each task (`struct task_info_s`) will be equipped with additional `3` elements of `uintptr_t` while globally we can have only `CONFIG_TLS_TASK_NELEM` number of destructors (`static tls_dtor_t g_tlsdtor[CONFIG_TLS_TASK_NELEM];`). So if I'm having `10` tasks the `3 * 10 * sizeof(uintptr_t)` will be additionally allocated for tasks and `3 * sizeof(tls_dtor_t)` for destructors. For example `task_A` call `task_tls_alloc()` one time and `task_B` call `task_tls_alloc()` two times. Then `task_A` exits and `task_tls_destruct` is called. The loop will iterate `3` times and globally `if ((g_tlsset & mask) != 0)` will always hit a `true` condition (one time for `task_A` and two times for `task_B`) and `dtor((void *)elem);` will be called `3` times passing 2 times `(void *)0` to `dtor()`. So the questions are next:
> 
> 1. Is it expected that `task_tls_destruct()` will call TLS destructors for other tasks?

Yes, this API set is designed to fix the global variables issue. For example, one library use the global variable `a`, 'b'.... And multiple service call the function exported by this library. So we need support multiple instance of `a`, 'b'... at the same time. One approach is:

1. Add xxx_initialize function to malloc and return a struct contain 'a', 'b'...
2. All function add one argument to accept the pointer to the memory returned in step 1
3. Use this struct instead of global variable
3. Add xxx_deinialize function to free the memory

The problem is that we need extend all pubic function accepted one instance pointer. It's hard to frequently-used library. Another approach (this PR) is:

1. library call task_tls_alloc to allocate one task index once in the whole lifetime
2. Call task_tls_get when need access global variable
3. If return NULL, allocate and initialize the global variables and call task_tls_set to save this pointer
4. Free the memory in task_tls_destruct callback when the task exit

The second approach don't need modify the public function prototype which is critical in some case.

> 2. If yes, then should we have a prerequisite that `dtor()` should check for `NULL` inside that call?

It expect the callback check the special value.

> 3. Is it expected that after `task_tls_destruct()` is called the while all `CONFIG_TLS_TASK_NELEM` are allocated the new `task_tls_alloc()` will always fail because `g_tlsset` never resets bits?

Yes, as the above example, task index is global to prepare the next task which may use the library function again.

> 4. What it `task_tls_destruct()` is called in the middle of `task_tls_alloc()`, for example after `g_tlsdtor[candidate] = dtor;` and before `ret = candidate;`? Will the `dtor` be ready to accept `NULL` (that actually comes back to 2)

Yes, the callback need handle this case carefully.

> 5. Is it expected to have `3 * (N - 1) * sizeof(uintptr_t)` (where `N` is the number of tasks in the system) of waste memory

This interface is designed to the common library(e.g. libuv):

1. Many processes use it
2. it use global variable internally

If the library is just used by one program, or multiple program at the different time, global variables is still the best choice.

Reply:
this change seems to be more `TGS` (task global storage) then `TLS`

Reply:
Maybe I'm missing some exact use case that you are trying to address, so posting a lot of question. If you can share use case then I can take a look from a different angle

Reply:
> Why do we need `ta_telem[CONFIG_TLS_TASK_NELEM];` to be a member of `struct task_info_s`? I mean why it can't be global? and have only allocated index bits in `struct task_info_s`

Because, each task need point to the different instance(please, consider my example, the library has the global variables, but want to support multiple clients at the same time):

1. All task(at the same index) use the same destructor to free the resource
2. Different task point to the different instance

If you compare to TLS:

1. All thread in the same task share the same destruct callback
2. Different thread point to the different instance

From the concept, task local storage is very similar to thread local storge, but at the task level.

1. Thread local storage allow the code allocate a task wide index with a destructor and attach a different pointer to the different thread at the same index in the same task.
2. Task local storage allow the code allocate a global index with a destructor and attach a different pointer to the different task at the same index.

Reply:
Yes. I will examine the use case more carefully early next week. I briefly went thru lines now and I have a feeling that I solved a similar problem with `pthread_once` + `pthread_key_create` on the library init and then the created key was used to access allocated data. I will try to understand what is the difference, but definitely need some more deep dive into what you are trying to solve.

Reply:
@xiaoxiang781216 Range limit added.

Reply:
> Yes. I will examine the use case more carefully early next week. I briefly went thru lines now and I have a feeling that I solved a similar problem with `pthread_once` + `pthread_key_create` on the library init and then the created key was used to access allocated data (please see example from https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_key_create.html). I will try to understand what is the difference, but definitely need some more deep dive into what you are trying to solve.

Your method can't handle multiple task case since pthread_key is specific to one task, in the new task you have to call pthread_key_create again. Basically, task local storage(attach pointer to each task) try to handle the similar problem as thread local storage(attach pointer to each thread in one task), but at the task level.

It isn't a problem in the traditional POSIX OS(e.g. Linux, BSD, NuttX' Kernel mode), since they utilize MMU and each task/process has separate virtual address space, but it's an essential combability issue in NuttX's FLAT/PROTECTED mode.
Apache NuttX Wiki has some article describe this problem:
https://cwiki.apache.org/confluence/display/NUTTX/Linux+Processes+vs+NuttX+Tasks
https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=152115079

Reply:
Now I see what you are trying to overcome. Current version is fine as an initial approach. We can enhance it later to overcome a global limitation of `CONFIG_TLS_TASK_NELEM` indexes per system.